test/hotspot/jtreg/runtime/cds/appcds/jcmd/JCmdTest.java

```diff
@@ -0,0 +1,364 @@
+                                               "LingeredApp source: shared objects file (top)",
+                                               "Hello source: shared objects file (top)"};
+
+    static String testJar = null;
+    static String bootJar = null;
+    static String allJars = null;
+
+    private static void buildJar() throws Exception {
+        testJar = JarBuilder.build("test", TEST_CLASSES);
+        bootJar = JarBuilder.build("boot", BOOT_CLASSES);
+        Path testJarPath = FileSystems.getDefault().getPath(testJar);
+        Path bootJarPath = FileSystems.getDefault().getPath(bootJar);
+        System.out.println("Jar file created: " + testJarPath.toString());
+        System.out.println("Jar file created: " + bootJarPath.toString());
+        allJars = testJar+ File.pathSeparator + bootJar;
+    }
+
+    private static boolean argsContain(String[] args, String flag) {
+         for (String s: args) {
+             if (s.contains(flag)) {
+                 return true;
+             }
+         }
+         return false;
+    }
+
+    private static boolean argsContainOpts(String[] args, String... opts) {
+        boolean allIn = true;
+        for (String f : opts) {
+            allIn &= argsContain(args, f);
+            if (!allIn) {
+                break;
+            }
+        }
+        return allIn;
+    }
+
+    private static LingeredApp createLingeredApp(String... args) throws Exception {
+        JCmdTestLingeredApp app  = new JCmdTestLingeredApp();
+        try {
+            LingeredApp.startAppExactJvmOpts(app, args);
+        } catch (Exception e) {
+            // Check flags used.
+            if (argsContainOpts(args, new String[] {"-Xshare:off", "-XX:+RecordDynamicDumpInfo"}) ||
+                argsContainOpts(args, new String[] {"-XX:+RecordDynamicDumpInfo", "-XX:ArchiveClassesAtExit="})) {
+                // app exit premature due to incompactible args
+                return null;
+            }
+            Process proc = app.getProcess();
+            if (e instanceof IOException && proc.exitValue() == 0) {
+                // Process started and exit normally.
+                return null;
+            }
+            throw e;
+        }
+        return app;
+    }
+
+    static int logFileCount = 0;
+    private static void runWithArchiveFile(String archiveName, boolean useBoot,  String... messages) throws Exception {
+        List<String> args = new ArrayList<String>();
+        if (useBoot) {
+            args.add("-Xbootclasspath/a:" + bootJar);
+        }
+        args.add("-cp");
+        if (useBoot) {
+            args.add(testJar);
+        } else {
+            args.add(allJars);
+        }
+        args.add("-Xshare:on");
+        args.add("-XX:SharedArchiveFile=" + archiveName);
+        args.add("-Xlog:class+load");
+
+        LingeredApp app = createLingeredApp(args.toArray(new String[0]));
```

Comment:
Instead of `new String[0]`, I think it is more intuitive to use `new String[args.size()]`.